### PR TITLE
Use warp size from `runtime.getDeviceProperties`

### LINKED ISSUE #6523: Use `warpSize` to get the warp size in ROCm environment
CuPy, currently, has a hard-coded warp size of 64 for ROCm environment. However, AMD gpu's warp size is not necessarily 64; for example, Navi21 that ROCm 5.0 supports has the warp size of 32. The following can be a good solution.

https://github.com/pytorch/pytorch/blob/2cefbb71cf65696e28ee1bdfce06d6846285e5fb/c10/macros/Macros.h#L322

Thanks @amathews-amd for pointing this out.

### PR DESCRIPTION
**Pull request by asi1024**
Close #6523
cc: @ajflood @amathews-amd @kmaehashi @leofang @takagi

### CODE REVIEW

File: cupy/_core/_kernel.pyx [hunk 1]
```diff
@@ -45,6 +45,12 @@ cdef inline bint _contains_zero(const shape_t& v) except? -1:
     return False
 
 
+@_util.memoize(for_each_device=True)
+def _get_warpsize():
```
Comment (asi1024) [inline cupy/_core/_kernel.pyx:49]: Should we provide this function to users as a public API?

Comment (kmaehashi) [reply cupy/_core/_kernel.pyx:49]: I think not needed as users can query this via existing API.

Btw, how about using `cupy.cuda.device._get_device().attributes["warpSize"]`?

Comment (asi1024) [reply cupy/_core/_kernel.pyx]: `Device.attributes["WarpSize"]` does not work in the CI environment, so I will revert the implementation to use `runtime.getDeviceProperties`.

```py
>>> from cupy.cuda import device
>>> device._get_device().attributes['WarpSize']
0
```

File: cupyx/jit/_interface.py [hunk 2]
```diff
@@ -182,8 +181,7 @@ def wrapper(func):
 blockIdx = _internal_types.Data('blockIdx', _cuda_types.dim3)
 gridDim = _internal_types.Data('gridDim', _cuda_types.dim3)
 
-warpsize = _internal_types.Data(
-    '64' if runtime.is_hip else '32', _cuda_types.uint32)
+warpsize = _internal_types.Data('warpSize', _cuda_types.uint32)
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupyx/jit/_interface.py:184]: This is a built-in variable!
https://docs.nvidia.com/cuda/cuda-c-programming-guide/#built-in-variables

File: cupyx/jit/_interface.py [hunk 3]
```diff
@@ -182,8 +181,7 @@ def wrapper(func):
...
 warpsize.__doc__ = r"""Returns the number of threads in a warp.
 
 In CUDA this is always 32, and in ROCm/HIP always 64.
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupyx/jit/_interface.py]: Removed this line in e785e71.

File: cupyx/jit/_interface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply cupyx/jit/_interface.py:184]: @leofang The built-in `warpSize` type is int32, but the current `cupyx.jit.warpsize` is uint32. Do yo think should we 1. fix the type to int32 to follow CUDA specifications, or 2. keep the type for backward compatibility and emit uint32 `(unsigned int)warpSize`?

[the remark this answers is not in the document]
Comment (leofang) [reply cupyx/jit/_interface.py:184]: I think we should just fix it. I don't think it'd be breaking any user code?

**Review by emcastillo — APPROVED** [review]

**Pull request merged**